### ISSUE TITLE
vim packages not being folded.

### ISSUE THREAD

**Issue by Jeansen** [vim-perl/vim-perl#251]
# Description of issue
I am not able to fold packages.

Your description here!

Maybe I am missing something, but though I have set `foldmethod=syntax` and fiddled around with all the different settings, I get everything to work - except code-folding for packages. See below:

```
let g:perl_fold = 1
let g:perl_include_pod = 1
let g:perl_extended_vars=1
let g:perl_fold_blocks = 1
```
I've also tried `let g:perl_nofold_packages` with 0/1, but no setting worked.

If it is a real bug, I'd definitely appreciate a fix :) If I just missed something, I would be grateful for any help.

**Comment by hoelzro** [comment]
Thanks for the bug report, @Jeansen! It seems that this form is folded:

```perl
package Null;

my $null = bless {}, __PACKAGE__;
sub AUTOLOAD {$null}
sub ok       {0}
```

But not the one you posted.  Let's see if we can get that fixed!

**Comment by Jeansen** [comment]
Hey @hoelzro! Thank's for the quick answer. I would really appreciate a fix :smile:  If I can be of support in any way, just let me know!

**Comment by hoelzro** [comment]
@Jeansen I just pushed some tests for "braced" packages - we could benefit from more tests (ex. `package NAME VERSION { ... }` is untested), and the implementation seems to have a bug with your exact example (for some reason the fold ends at `sub ok`).  Also nice would be to add a CPAN module that uses brace packages to the corpus test, but that's something I'll probably need to do.

If you feel comfortable writing tests, that would be great - if you feel up to the challenge of debugging why the folding Vimscript isn't working, that would be even better!

**Comment by Jeansen** [comment]
@hoelzro Thank you. I'll have a look and see, what I can do. I am no Vimscript expert, but I'll take the challenge. But it might take some time.

**Comment by Jeansen** [comment]
@hoelzro  I am not sure, if I am missing something, but when I use the following vimrc and try to fold the snippet from https://github.com/vim-perl/vim-perl/issues/251#issuecomment-340282312 then I do not see any change. Here's my minimalistic vimrc:

```
#!/usr/bin/vim

set encoding=utf-8
scriptencoding utf-8
set termencoding=utf-8
set shell=bash
filetype plugin indent on 
syntax on
set foldmethod=syntax

let g:perl_fold = 1
let g:perl_nofold_packages=0

call plug#begin('~/.vim/plugged')

Plug 'vim-perl/vim-perl', { 'branch': 'dev', 'for': 'perl', 'do': 'make clean carp dancer highlight-all-pragmas moose test-more try-tiny' }

call plug#end()
```

**Comment by hoelzro** [comment]
Hmm...looking at the code it turns out that setting `perl_nofold_packages` to *anything* will disable the folding; I'll fix that up in a bit!

**Comment by Jeansen** [comment]
Oh...well. Now I am ashamed to not have checked it myself :blush:  Must be, because I am still in the Perl-Universe :laughing:

**Comment by Jeansen** [comment]
And yes, without `perl_nofold_packages` it works as described

**Comment by Jeansen** [comment]
I have made a PR with something working. Could you check, please? :smile:

**Comment by Jeansen** [comment]
BTW, I noticed, that I have to delete my perl.vim in the vim home folder. Should it not be overwritten/overruled  by the plugin installation?

**Comment by hoelzro** [comment]
It depends on how you're installing your plugins - if you're using something like Pathogen, for example, `$HOME/.vim/syntax/perl.vim` will take precedence over `$HOME/.vim/bundle/vim-perl/syntax/perl.vim`.

**Comment by hoelzro** [comment]
@Jeansen Sorry for the delay on this - I don't get a lot of time to work on vim-perl, and every time I do I feel like I'm just remembering where I was at!  I believe I have fixed the package folding issue - could you grab what's on the dev branch and verify that it works?

**Comment by Jeansen** [comment]
@hoelzro No problem, same here ;) Backage folding issue seems to be resolved. Great work! Thank you :smiley:

**Comment by Jeansen** [comment]
Hey @hoelzro. I hope you are doing fine? After some months of absence, I finally found some time to get back to my project. Thou your fix solve some folding problems, it still is not doing everything as expected. It seems to be a bit too eager ;-)

If I use your current fix from the development branch and use syntax folding on https://github.com/Jeansen/cdmn/blob/master/cdmn , then some packages are recursively folded beneath  package `cdmn::math`, though they are defined on the same level. That is,`cdmn::contstants` is folded beneath package `cdmn::math`. `cdmn::proc` is folded beneath `cdmn::constants` and so son.

Could you check again? I bet it is just a very little change that has to be done ;)

**Comment by hoelzro** [comment]
Hi @Jeansen! I will try to check this out later today

**Comment by hoelzro** [comment]
Hi again @Jeansen - sorry this took me so long to check out!  I had some time to look tonight, and I was able to fix it - but not without breaking folding for anonymous subs.  I think I'll need to step back and see how I unify these two ideas.

**Comment by Jeansen** [comment]
Hi @hoelzro. Thanks for all your efforts! Do not hurry yourself. There is no rush :wink:  But maybe you will find  some more time to also fix the folding for anonymous subs. Keep on the good work! :smile:

**Comment by Jeansen** [comment]
@hoelzro: Did you push your fix? I cannot find it on dev ...

**Comment by hoelzro** [comment]
@Jeansen No fix yet, I'm afraid - I can't push what I have since it breaks anonymous subs =(

**Comment by Jeansen** [comment]
Oh, ok. No problem. Then I'll wait until you find a fix :smiley:

**Comment by sdondley** [comment]
Has this issue been abandoned? At first glance, this plugin seems unmaintained and kind of unusable  but maybe I'm using it wrong.

**Comment by petdance** [comment]
Not abandoned. Still maintained. But it looks like nobody has a fix for it.

**Comment by sdondley** [comment]
K, thanks. I'm investigating. Started with a clean vim install on a new OS. I'm poking around. I just know enough vim scripting to be dangerous, though.

**Comment by petdance** [comment]
That's OK.  **Any** vim scripting is dangerous.

**Comment by sdondley** [comment]
Well, so commenting out line 394 in `/syntax/perl.vim` restores indenting but you lose highlighting.

`syn match perlFunction "\<sub\>\_s*" nextgroup=perlSubDeclaration`

**Comment by sdondley** [comment]
And commenting out line 393 keeps both the highlighting and the indenting. But not sure if that might break folding or what. Line 393:

`syn region perlSubDeclaration start="" end="[;{]" contains=perlSubName,perlSubPrototype,perlSubAttribute,perlSubSignature,perlComment contained transparent`

**Comment by sdondley** [comment]
Yeah, that breaks folding. It folds all subs into a singe fold.

**Comment by sdondley** [comment]
OK, this seems to do the trick:

`syn region perlSubDeclaration start="\_.\{-}{"hs=e+1 end="[;{]"he=e-1 contains=perlSubName,perlSubPrototype,perlSubAttribute,perlSubSignature,perlComment contained transparent`

This line should replace line 392 (or thereabout).

Untested, though. How do I do that? Is this a perl module I can download?

UPDATE: NVM, figured out how to run the tests.

**Comment by hoelzro** [comment]
@sdondley Thanks for looking into this - by the way, if you want quick feedback on whether or not a change would break something, the test suite runs highlighting and folding against some known sharp corners, as well as a pretty good corpus of Perl modules from the CPAN.  It even gives you pretty feedback like this for highlighting failures:

![highlight-test-example](https://user-images.githubusercontent.com/107804/55452831-f4194000-559e-11e9-84b2-3505b07f6506.png)

You may need to install some Perl modules to run the test suite, but we're happy to assist!

**Comment by sdondley** [comment]
Yeah, just got done running the tests. As I suspected, it failed on anon subroutines. Not sure what the corpus tests are but it failed many of those.

But basically the issue is that the regions is "covering" the braces. By adjusting where the region begins and ends with the offset values, you can "move" the beginning and ending of the region so the braces are "uncovered." Then the indentation will work properly.

**Comment by sdondley** [comment]
Oops, all these previous comments of mine were mean for https://github.com/vim-perl/vim-perl/issues/267. Sorry!

**Comment by hoelzro** [comment]
Ah, ok - vim-perl could really use some proper indent tests!  The "corpus" tests are essentially golden master tests that make sure that highlighting and folding is consistent across a few modules I picked - they make sure that, for the most part, users don't see code suddenly highlight differently when a change is made.
